src/GazeboThruster.cpp

```diff
@@ -13,48 +12,96 @@ GazeboThruster::~GazeboThruster()
+    {
+        std::string linkName = (*link)->GetName();
+        std::size_t found = linkName.find("underwater_thruster_body");
+        if(found != std::string::npos )
+        {
+            gzmsg <<"GazeboThruster: found link: " << linkName << std::endl;
+            linkName.erase( found - 2 );
+            gzmsg <<"GazeboThruster: JointState element name in ROCK joint input port must match the link name: "
+                    << linkName << std::endl;
+            thrusterInput.insert( std::make_pair( linkName, 0.0 ) );
+        }
+    }
+    if(thrusterInput.empty())
+        gzthrow("GazeboThruster: thruster link name underwater_thruster_body not found.");
 }
 
-void GazeboThruster::readInput(ThrusterInput &_thruster_input)
+
+void GazeboThruster::initNode()
 {
-    thruster_input = _thruster_input->frequency();
+    // Initialize communication node and subscribe to gazebo topic
+    node = transport::NodePtr(new transport::Node());
+    node->Init();
+    std::string topicName = model->GetName() + "/thruster";
+    thrusterSubscriber = node->Subscribe("~/" + topicName,&GazeboThruster::readInput,this);
+    gzmsg <<"GazeboThruster: create gazebo topic /gazebo/"+ model->GetWorld()->GetName()
+            + "/" + topicName << std::endl;
 }
 
 
-double GazeboThruster::thrusterMathModel()
+void GazeboThruster::readInput(JointsMSG& jointsMSG)
 {
-    // Linear model
-    double a = 1.0, b = 0.0;
-    double thruster_output = a* thruster_input + b;
-    return thruster_output;
+    // Read buffers and update joints data
+    for(int i = 0; i < jointsMSG->raw_size(); ++i)
+    {
+        const rock_thruster::msgs::Raw& jointRaw = jointsMSG->raw(i);
+        for(GazeboJoint::iterator joint = thrusterInput.begin();
+                joint != thrusterInput.end(); ++joint)
+        {
+            if(joint->first == jointRaw.name() )
+            {
+                joint->second = thrusterMathModel( jointRaw.raw() );
+            }
+        }
+    }
+
+    for( int i = 0; i < jointsMSG->effort_size(); ++i)
+    {
+        const rock_thruster::msgs::Effort& jointEffort = jointsMSG->effort(i);
+        for(GazeboJoint::iterator joint = thrusterInput.begin();
+                joint != thrusterInput.end(); ++joint)
+        {
+            if(joint->first == jointEffort.name() )
+                joint->second = jointEffort.effort();
+        }
+    }
 }
 
-void GazeboThruster::applyForce(double force)
+double GazeboThruster::thrusterMathModel(double input)
 {
-    link->AddForce( math::Vector3(force,0,0) );
+    // Linear model
+    double a = 1.0, b = 0.0;
+    return a*( input ) + b;
 }
 
 
-
+void GazeboThruster::updateBegin(common::UpdateInfo const& info)
+{
+    for(GazeboJoint::iterator input = thrusterInput.begin();
+            input != thrusterInput.end(); ++input)
+    {
+        physics::LinkPtr link = model->GetLink(input->first + "::underwater_thruster_body");
```

Comment:
I don't think we can switch gazebo link name order.
Gazebo create links this way:  **THRUSTER_NAME::underwater_thruster_body**
Where the user define **THRUSTER_NAME** in world file.
And I need the full link name to access them. 

The general idea to update the link effort is as fallows: 
  - the thruster rock component send the link name and effort using protobuf msgs;
  - then the plugin check the incoming link names to update their efforts;

As all thruster link names have  **underwater_thruster_body** in their names I didn't want the user to send the whole link name **THRUSTER_NAME::underwater_thruster_body**.
I can rename **underwater_thruster_body** to **thruster**, if you wish. 
But it would be simple to avoid that.